azure/scope/workload_identity.go

```diff
@@ -0,0 +1,157 @@
+|                           | application or user-assigned managed identity.        |
+|AZURE_TENANT_ID            | The tenant ID of the Azure subscription.              |
+|AZURE_FEDERATED_TOKEN_FILE | The path of the projected service account token file. |
+|-----------------------------------------------------------------------------------|
+
+In addition to the service account, it also projects a signed service account token to the
+workload's volume(in this case the capz pod). The volume name is `azure-identity-token`
+which is mounted at path `/var/run/secrets/azure/tokens/azure-identity-token` to the pod.
+
+*/
+
+const (
+	// AzureFedratedTokenFileEnvKey is the env key for AZURE_FEDERATED_TOKEN_FILE.
+	AzureFedratedTokenFileEnvKey = "AZURE_FEDERATED_TOKEN_FILE"
+	// AzureClientIDEnvKey is the env key for AZURE_CLIENT_ID.
+	AzureClientIDEnvKey = "AZURE_CLIENT_ID"
+	// AzureTenantIDEnvKey is the env key for AZURE_TENANT_ID.
+	AzureTenantIDEnvKey = "AZURE_TENANT_ID"
+	// AzureTokenFilePath is the path of the projected token.
+	AzureTokenFilePath = "/var/run/secrets/azure/tokens/azure-identity-token" // #nosec G101
+)
+
+type workloadIdentityCredential struct {
+	assertion string
+	file      string
+	cred      *azidentity.ClientAssertionCredential
+	lastRead  time.Time
+}
+
+// WorkloadIdentityCredentialOptions contains the configurable options for azwi.
+type WorkloadIdentityCredentialOptions struct {
+	azcore.ClientOptions
+	ClientID      string
+	TenantID      string
+	TokenFilePath string
+}
+
+// NewWorkloadIdentityCredentialOptions returns an empty instance of WorkloadIdentityCredentialOptions.
+func NewWorkloadIdentityCredentialOptions() *WorkloadIdentityCredentialOptions {
+	return &WorkloadIdentityCredentialOptions{}
+}
+
+// WithClientID sets client ID to WorkloadIdentityCredentialOptions.
+func (w *WorkloadIdentityCredentialOptions) WithClientID(clientID string) *WorkloadIdentityCredentialOptions {
+	w.ClientID = clientID
+	return w
+}
+
+// WithTenantID sets tenant ID to WorkloadIdentityCredentialOptions.
+func (w *WorkloadIdentityCredentialOptions) WithTenantID(tenantID string) *WorkloadIdentityCredentialOptions {
+	w.TenantID = tenantID
+	return w
+}
+
+// GetProjectedTokenPath return projected token file path from the env variable.
+func GetProjectedTokenPath() (string, error) {
+	tokenPath := os.Getenv(AzureFedratedTokenFileEnvKey)
+	if strings.TrimSpace(tokenPath) == "" {
+		return AzureTokenFilePath, nil
+	}
+	return tokenPath, nil
+}
+
+// WithDefaults sets token file path. It also sets the client tenant ID from injected env in
+// case empty values are passed.
+func (w *WorkloadIdentityCredentialOptions) WithDefaults() (*WorkloadIdentityCredentialOptions, error) {
+	tokenFilePath, err := GetProjectedTokenPath()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get token file path for identity")
+	}
+	w.TokenFilePath = tokenFilePath
+
+	// Fallback to using client ID from env variable if not set.
```

Comment:
Is this the workaround for #3409?

I'm looking at that but not sure what needs to be fixed (yet).

Reply:
No.
It is still allowed to leave client ID and tenant ID fields on AzureClusterIdentity object to be empty. So this is fallback for cases where those fields are not set on AzureClusterIdentity. 

So the behaviour is like this: 
- CAPZ uses the tenant Id and the client ID as specified on AzureClusterIdentity. 
- If client ID is not specified, it tries to read from the env.
- If tenant ID is not specified, it tries to read from the env.

Reply:
Is it not possible to make client ID and tenant ID mandatory if identity type is `WorkloadIdentity`? IMO having fallbacks to env vars could cause confusion and requires additional documentation on the behavior when both CRD and env vars are configured.

Reply:
Couple of thing here:
1. AzureClusterIdentity is referenced by AzureCluster.
2. So if I want to create a workload cluster by using/referencing a AzureClusterIdentity capz should always use the IDs from it.

Also, I do not know why client ID and tenant ID are optionally required and not strictly required on AzureClusterIdentity. 

So I am not sure if we should make a special case for workload identity, though it is possible to do so.

Having said that, if I leave the workload identity discussion aside, then still it makes more sense to use IDs present on AzureClusterIdentity to be used. (e.g ManualServicePrincipal )

I am also thinking about the following in deciding the priority:
1. Read IDs from AzureClusterIdentity and fall back to read ID from env if not present on it. Or
2. Read IDs from env and fall back to using IDs present on AzureClusterIdentity.

Here [2] looks semantically confusing/incorrect to me.

Reply:
cc @CecileRobertMichon

Reply:
I think it would make sense to make client ID and tenant ID required, AFAIK there is not Identity type that doesn't need them. Can we create an issue and do this as a follow up?